[X86_64] Fix empty field error in vaarg of C++.

CoTinker opened the pull request:
Such struct types:
```
struct {
  struct{} a;
  long long b;
};

stuct {
  struct{} a;
  double b;
};
```
For such structures, Lo is NoClass and Hi is Integer/SSE. And when this structure argument is passed, the high part is passed at offset 8 in memory. So we should do special handling for these types in EmitVAArg.Fix https://github.com/llvm/llvm-project/issues/79790 and fix https://github.com/llvm/llvm-project/issues/86371.

CoTinker committed:
[X86_64] Fix empty field error in vaarg of C++.

Such struct types:
```
struct {
  struct{} a;
  long long b;
};

stuct {
  struct{} a;
  double b;
};
```
For such structures, Lo is NoClass and Hi is Integer/SSE. And when
this structure argument is passed, the high part is passed at
offset 8 in memory. So we should do special handling for these types
in EmitVAArg.

CoTinker commented:
Ping~

efriedma-quic commented on clang/lib/CodeGen/Targets/X86.cpp:
```diff
@@ -3124,26 +3124,40 @@ RValue X86_64ABIInfo::EmitVAArg(CodeGenFunction &CGF, Address VAListAddr,
     // Copy into a temporary if the type is more aligned than the
     // register save area.
-    if (TyAlign.getQuantity() > 8) {
+    if (neededInt && TyAlign.getQuantity() > 8) {
```
Can we refactor the code so it's clear we never create two temporaries for the same va_arg?  Use an else-if instead of a separate if.

It looks it's possible that we end up reading past the end of the register save area for something like:

```
struct {
  long long b;
  struct{} a;
};
```

Can you also fix that while you're here?  (Not sure that actually has any visible effect, but better to be on the safe side.)

CoTinker replied:
I am glad to fix it.

CoTinker committed:
[X86_64] Fix edge cases

Fix edge cases that may reading past the end of the register save area.
Such as:
```
struct {
  long long a;
  struct {} b;
};
```

efriedma-quic commented on clang/lib/CodeGen/Targets/X86.cpp:
````diff
@@ -3124,26 +3124,76 @@ RValue X86_64ABIInfo::EmitVAArg(CodeGenFunction &CGF, Address VAListAddr,
+    //   ```
+    //   struct {
+    //     long long a;
+    //     int b;
````
This example doesn't seem right; the size of the register save area has to be a multiple of 8.

Given that, we can simplify the "RegSize" to just `neededInt ? neededInt * 8 : 16`, I think?

CoTinker replied:
Yeah, you're right.

CoTinker replied:
SSE size is 16?
Consider such structure：
```
struct {
  double a;
  struct {} b;
}
```
Maybe `RegSize = neededInt ? neededInt * 8 : 8`, I think?

efriedma-quic replied:
The relevant "size" here is the size of the slots in the register save area.  Which is 16 bytes for fp registers.

CoTinker committed:
fix testcase

efriedma-quic reviewed (approved):
LGTM

CoTinker commented:
Thanks.
